Rendering fix

kamil-matula opened the pull request:
Scrolling down the list will result in spotting cards with placeholders - these pokemons has not been rendered yet.

![image](https://user-images.githubusercontent.com/43967269/118524577-96c0b880-b73e-11eb-9d9b-371d7d46a694.png)

kamil-matula committed:
Divided Data.js into 2 files

Moved Data.js from main directory to contexts and divided into PokeApiData.js and PokeJSON.js to separate building JSON object from sending requests to API. Also added constant urls.

kamil-matula committed:
Added displaying all pokemons from API

DONE:
1. Added rendering cards with basic JSONs - just ID, name and pokeball-placeholder.
2. Moved fetching pokemon's details to PokemonCard.js.
3. Added replacing basic JSONs with full JSONs to render expected card.
4. Fixed sorting (by adding default state for filters) and resetting current pokemon.

TODO:
1. Find better way to render pokemon cards (it takes too long).
2. Fix sorting/filtering pokemons - if the list (898 pokemons!) hasn't rendered yet, it will not be full and it will not update.
3. Find a way to cache the data, so further sessions of using application will be much faster.

Blazevarjo commented on src/components/PokemonCard.js:
```diff
@@ -1,16 +1,22 @@
 
 export default function PokemonCard({ onPress, id, color, name, sprite, types }) {
   const colors = useTheme().colors;
-  const cardColor = colors.pokemon.background[color];
+
+  const { updatePokemonObject } = useContext(PokemonDataContext);
+  useEffect(() => {
+    pokemonByNameOrNumber(id).then((pokemon) => updatePokemonObject(pokemon));
+  }, []);
 
```
Should be pure component so any effects or functionality shouldn't appear here

Blazevarjo commented on src/components/PokemonCard.js:
```diff
@@ -1,16 +1,22 @@
+  }, []);
 
   return (
-    <Surface style={[styles.container, { backgroundColor: cardColor }]}>
+    <Surface style={[styles.container, { backgroundColor: colors.pokemon.background[color] }]}>
       <TouchableRipple borderless={true} style={styles.touch} onPress={() => onPress()}>
```
It's better to pass prop with color

Blazevarjo commented on src/screens/main/Home.js:
```diff
@@ -10,11 +10,13 @@ export default function Home() {
-    ({ item: { id, name, height, weight, sprite, types, color } }) => (
+    ({ item: { id, name, height, weight, sprite, types, color, loaded } }) => (
       <PokemonCard
         onPress={() => {
-          navigation.navigate('Details');
-          updateCurrentPokemonId(id);
+          if (loaded == true) {
+            navigation.navigate('Details');
+            updateCurrentPokemonId(id);
+          }
         }}
         id={id}
```
If statement in render item also reduce performance in list rendering

Blazevarjo replied:
Also 
```
if (loaded==true)
```
should be changed to 
```
if(loaded)
```

Blazevarjo commented on src/contexts/PokemonDataContext.js:
```diff
@@ -36,7 +36,13 @@ export const PokemonDataProvider = ({ children }) => {
-  const resetCurrentPokemon = () => setCurrentPokemon(null);
+  const resetCurrentPokemon = () => setCurrentPokemonId(null);
+
+  const updatePokemonObject = (newPokemonObject) => {
+    let index = pokemons.indexOf(pokemons.find((pokemon) => pokemon.id == newPokemonObject.id));
+    pokemons[index] = newPokemonObject;
+    // TO DO: refresh item
```
State should be treated as immutable.

That's cleaner approach
```
pokemons.findIndex((pokemon) => pokemon.id == newPokemonObject.id)
```

Blazevarjo reviewed (changes requested):
I think that this approach  affects poor performance of flatlist. Better way to tweak performance would be to apply Pure Components to pokemon card, cache data and change fetching whole data to smaller batches. We can also think about optimizing images, but first we should focus on developing major functionality, then will be time for optimizing.

In my opinion we should abondon this PR at least for a while

kamil-matula committed:
Refactored rendering to use batches

DONE:
1. Moved updating pokemon's JSON from PokemonCard.js to PokemonDataContext.js.
2. Added remaining default states.
3. Added useState for id of currently rendered pokemon.
4. Refactored updatePokemonObject function.
5. Added useEffect which updates multiple JSON's at the same time (in other words: uses "batches").

TODO:
1. Merge main into this branch and clean up.
2. Fix sorting/filtering pokemons - if the list (898 pokemons!) hasn't rendered yet, it will not be full and it will not update.
3. Find a way to cache the data, so further sessions of using application will be much faster.
4. Eliminate "VirtualizedList: You have a large list that is slow to update" warning.

kamil-matula committed:
Merge branch 'main' into RenderingFix

kamil-matula committed:
Cleaned up after merge

kamil-matula committed:
Refactored re-rendering chain

1. Moved LoadFavouritesIds and SaveFavouritesIds to new file: StorageActions.js.
2. Added some new comments.
3. Added filtering pokemons to re-rendering chain.

kamil-matula committed:
Changed some parameters

kamil-matula committed:
Prepared three variants of rendering (for testing purposes)

There are three variants of rendering:
1. Old variant, which uses constant amount value (and too many promises).
2. New variant, which uses placeholders and re-renders pokemon cards batch by batch (probably it would be better if the objects will be updated by whole batches, not by mapping, but, because of sorting whole array, it is hard to do this).
3. Mixed variant, which checks how many pokemons is in API and then uses old variant to render everything at one time.
